### PR TITLE
Add policy CRDs

### PR DESCRIPTION
**Pull request by olix0r**
We've implemented new CRDs to support configuration and discovery of
inbound policies. This change imports these CRDs from
linkerd/polixy@25af9b5 and adds them to the default Linkerd installation:

- `servers.policy.linkerd.io`
- `serverauthorizations.policy.linkerd.io`

### CODE REVIEW

**Review by adleong — APPROVED** [review]
Verified that installing Linkerd is successful and adds these CRDs and that uninstalling Linkerd removes them.

I'm curious about the commented out additional printer columns.  Would these be useful to uncomment?

**Comment by olix0r** [comment]
> I'm curious about the commented out additional printer columns. Would these be useful to uncomment?

Last I checked, the current state of these was somewhere between ugly and incorrect. I agree it would be useful to revisit these, perhaps before merging?

**Comment by olix0r** [comment]
@adleong 

> I'm curious about the commented out additional printer columns. Would these be useful to uncomment?

Last I checked, the current state of these was somewhere between ugly and incorrect. I agree it would be useful to revisit these, perhaps before merging?

**Comment by alpeb** [comment]
If these CRDs are gonna have default instances for the control plane pods deployed at install time, then we need to place the CRD definitions into a new `charts/linkerd2/crds` directory for the helm install method to install the defs before the instances:
https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you

And for the CLI to work we'll have to introduce some mechanism that blocks on the availability of those APIs before trying to persist its instances. I'd be happy to work on that if needed :+1:

File: charts/linkerd2/templates/policy-crd.yaml [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+                      type: array
+                      items:
+                        type: object
+                        required: [key, operator, values]
+                        properties:
+                          key:
+                            type: string
+                          operator:
+                            type: string
+                            enum: [In, NotIn]
```
Comment (alpeb) [inline charts/linkerd2/templates/policy-crd.yaml:61]: How about including `Exists` and `DoesNotExist` here as well (in which case `values` wouldn't be required)? (same for serverauth below)

File: charts/linkerd2/templates/policy-crd.yaml [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+              required:
+                - podSelector
+                - port
+              properties:
+
+                podSelector:
+                  type: object
+                  description: >-
+                    Selects pods in the same namespace.
+                  properties:
```
Comment (alpeb) [inline charts/linkerd2/templates/policy-crd.yaml:47]: Should we have `oneof: required[matchLabels], required[matchExpression]` like we do below for serverauth?

File: charts/linkerd2/templates/policy-crd.yaml [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+                            from clients that do not yet have a certificate.
+
+                        identities:
+                          description: >-
+                            Authorizes clients with the provided proxy identity
+                            strings (as provided via MTLS)
+
+                            The `*` prefix can be used to match all identities in
+                            a domain. An identity string of `*` indicates that
+                            all authentication clients are authorized.
```
Comment (alpeb) [inline charts/linkerd2/templates/policy-crd.yaml:238]: Wondering why is this handled in separate fields (`identities` and `suffixes`) in `inbound.proto`?

File: charts/linkerd2/templates/policy-crd.yaml (hunk 2 above)
Comment (olix0r) [reply charts/linkerd2/templates/policy-crd.yaml:47]: Yes, I think so.

File: charts/linkerd2/templates/policy-crd.yaml (hunk 1 above)
Comment (olix0r) [reply charts/linkerd2/templates/policy-crd.yaml:61]: If that's supported by Kubernetes in other contexts, sounds good.

File: charts/linkerd2/templates/policy-crd.yaml (hunk 3 above)
Comment (olix0r) [reply charts/linkerd2/templates/policy-crd.yaml:238]: The intent is to make the user-facing CRD configuration to be ergonomic: as a user, I just want to list a set of patterns.

But, for the internal proto API, we want the proxy to be crystal clear about what it's being asked to do (exact match vs suffix match). It's the job of the controller to do validation (i.e. ultimately as an admission controller for these resources, though that's not yet been implemented). I definitely do *not* want the proxy to have to deal with ambiguous configuration (parsing strings, etc).

File: charts/linkerd2/templates/policy-crd.yaml [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+                            properties:
+                              name:
+                                description: The ServiceAccount's name.
+                                type: string
+                                pattern: '^[a-z0-9]([-a-z0-9]*[a-z0-9])?$'
+
+                              namespace:
+                                description: >-
+                                  The ServiceAccount's namespace. If unset, the
+                                  authorization's namespace is usd.
```
Comment (mateiidavid) [inline charts/linkerd2/templates/policy-crd.yaml:263]:
```suggestion
                                  authorization's namespace is used.
```

**Review by mateiidavid — APPROVED** [review]

**Comment by alpeb** [comment]
Having dug deeper into Helm doc's I realized putting the CRD under the `crds` directory means `helm upgrade` won't touch it, and any upgrades for it will have to be done manually by the user.

Keeping it under `templates` doesn't have that problem. The docs say this has the downside that `helm delete` (or `helm rollback` to a previous version that doesn't have the CRD), by deleting the CRD, would also delete all its instances, which might be surprising for users. I think in our particular case that wouldn't be really surprising and should be fine.

If we go with the `templates` approach, we need then to solve the problem of how to create the initial CRD instances (for the kubelet to successfully communicate with pods for example). I can see three potential different approaches:

1) Like the CLI install stages, have one Helm chart for installing cluster-level resources and another one for the linkerd-namespace level resources. Just a little annoying, but has the additional advantage that the latter chart would only require namespace-level RBAC.

2) Install the initial CRD instances through a post-install hook. Would have to experiment more how this turns out. One surprising behavior is that `helm install --wait` will fail, because that will check the resources are ready _before_ the hook triggers. And without the CRD instances to allow the kubelet to perform its checks, they won't show ready. This might be a minor issue that we'd have to clearly document.

3) Don't rely on initial CRD instances and instead hard-code those cases in the controller.

**Comment by olix0r** [comment]
> If we go with the templates approach, we need then to solve the problem of how to create the initial CRD instances (for the kubelet to successfully communicate with pods for example).

Currently, there's no need to add policies to permit Kubelet traffic -- Kubelet traffic is _always_ permitted to _all_ ports. We may revisit this in the future, but it's the safest default. (Kubelet is necessarily a privileged operator--it starts and stops the containers, after all--so it's really a matter of whether we can trust the pod network model to not give Kubelet IPs to other pods. I think this reasonable for now).

Backing up for a minute: When would we need policy CRDs for the core control plane? (Yes, we _should_ provide them, but when are they _required_?)

The core control plane's proxies will discover their policies at startup (which also means that the identity controller probably needs to run its own policy controller for itself so it doesn't depend on an external controller to startup). Each policy controller has a [default-allow](https://github.com/linkerd/linkerd2/blob/8c7e7cec22e5c10d88718f82f334a95f3673199c/policy-controller/src/main.rs#L31-L32) policy -- that is, if there are no server/policy resources, what policy should the controller return? There are a few options here:

https://github.com/linkerd/linkerd2/blob/8c7e7cec22e5c10d88718f82f334a95f3673199c/policy-controller/k8s/index/src/default_allow.rs#L10-L17

If we run the core control plane with a policy other than `deny`, there's actually no need to write explicit policies for the core control plane... These would only be necessary if we wanted to have a default-deny policy.

Though, for the identity controller in particular, we probably want another type of default policy that does not actually require an authenticated client...

https://github.com/linkerd/linkerd2/blob/8c7e7cec22e5c10d88718f82f334a95f3673199c/policy-controller/core/src/lib.rs#L64-L65

We probably need a default policy that support the `TlsUnauthenticated` case so that the identity controller can accept TLS'd connections that do not yet have a client identity...

---

So, with these defaults in mind, we probably don't actually have a blocker on adding servers/policies for the core control plane.

That said, I'm really curious about the muiti-chart system you describe. For one, it couldn't really work for the CLI install... we'd need separate commands to install the cluster objects vs the controllers (this may be okay, but is a tradeoff worth considering). Is there any prior art for this? Are other projects splitting their cluster-level resources into separate helm charts? Is this going to smell weird to users?

**Comment by alpeb** [comment]
All that makes sense to me, thanks for the detailed explanation :+1: 

You're right about the CLI install, this can't be done in a single step. Our current multi-stage workflow would then become the norm.

As for having two charts, that's one of the two best practices described in the [helm docs](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-2-separate-charts) (the other one being the `crds` directory), so that should be fine. For example [istio](https://istio.io/latest/docs/setup/install/helm/) and [Datadog](https://artifacthub.io/packages/helm/datadog/datadog-crds) do that.

**Review by alpeb — APPROVED** [review]

**Pull request merged**